### ISSUE TITLE
[turborepo] turbo-ignore marks all apps as 'affected' on any lockfile change

### ISSUE THREAD

**Issue by neefrehman** [vercel/turbo#6030]
### What version of Turborepo are you using?

1.10.14

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Describe the Bug

Related to https://github.com/vercel/turbo/issues/2162

## TL;DR
Running a workspace-filtered `turbo-ignore`, after any lockfile change, results in that workspace being marked as 'affected', even when the lockfile change doesn't impact the workspace. 

## Full description
We have a monorepo that deploys multiple apps to Vercel, and have package updates managed by Renovate. With this setup, we frequently see package updates happening that should only affect one of the apps. For example, lets say that one app, "app_2", has `is-even` installed, and no other workspace in the repo does.

When `is-even` has its version updated by Renovate, preview deployments of every single one of our apps will be deployed to Vercel. When that PR is merged into the main branch, the same happens for production deployments. Here is an example of what the Vercel deployment logs for a different app say, about the `turbo-ignore` step for these changes:

```shell
≫   Using Turborepo to determine if this project is affected by the commit...
≫   Inferred "app_1" as workspace from "package.json"
≫   Using "build" as the task from the arguments
≫   Previous deployment ("a70f8196fc252817dc04dccd53cb5e25d5471834") for "app_1" on branch "renovate/is-even-4.x" is unreachable.
≫   Falling back to ref HEAD^
≫   Analyzing results of `npx turbo run "build" --filter=app_1...[HEAD^] --dry=json`
≫   This commit affects "app_1"
```

The rest of these builds happen fairly quickly, as every step in the pipeline is a `cache hit`, but ~1 minute is still taken for each one, mostly taken up by the `Deploying outputs` step. This slows down our build queue by quite a lot.

### Expected Behavior

My understanding is that `turbo-ignore`'s behaviour should match the same logic that `turbo` uses to determine if there is a cache hit or not. Based on https://github.com/vercel/turbo/issues/2162, and my observations above, I believe that lockfile updates should only trigger builds—via `turbo-ignore`—if the workspace in question is affected by the changed dependencies. In the example above, only "app_2" should have a deployment triggered, in preview or production.

### To Reproduce

1. Clone the [minimal starter](https://github.com/vercel/turbo/tree/main/examples/basic)
2. Install dependencies with pnpm
3. Update the lockfile in a way that shouldn't affect one of the apps (e.g. adding a `devDependency` to the root `package.json`)
4. Run `cd apps/web && npx turbo-ignore`
5. See that `turbo-ignore` now results in `This commit affects "web" ✓ Proceeding with deployment`

**Comment by tknickman** [comment]
The issue here looks to be with the specific clone that turbo-ignore is running against. In the output shared above:

> ≫   Previous deployment ("a70f8196fc252817dc04dccd53cb5e25d5471834") for "app_1" on branch "renovate/is-even-4.x" is unreachable.

This means that `turbo-ignore` is looking for the previously deployed commit for the branch, but it can't locate it for the comparison to determine what has changed. This can happen if a force push removes this commit in question, or (on Vercel) the commit was too far back to be pulled for the shallow clone. 

Does this happen for all renovate PR's?

**Comment by neefrehman** [comment]
@tknickman thanks for your comment!

Yes this happens for all Renovate PRs. The first deployment is new to the branch, and subsequent rebases that Renovate does are via force pushes. Though, my suspicion is that this is unrelated to the issue (although I see that I used a confusing example log, sorry!).

The exact ignore command we are using is `npx turbo-ignore --fallback=HEAD^`, so turbo is falling back to the `main` branch in these cases, and still doesn't determine that the commit should be ignored. This is also happening on the production deployment after the PR is merged to main, where the previous deployment _is_ found, and a build still happens even though it shouldn't.

Following the reproduction steps should also show you that this issue occurs after lockfile changes to the same branch.

**Comment by nathanhammond** [comment]
Hey @chris-olszewski! This is _supposed_ to be addressed by #3250, so I'm wondering if we have a regression in the SCM migration to Rust? (That's my best guess as to why we are failing to find the thing.)

**Comment by nathanhammond** [comment]
Reproduction as a script:
```
pnpm dlx create-turbo@canary test-lockfile-invalidation --package-manager pnpm
cd test-lockfile-invalidation/apps/web
pnpm add left-pad
git add --all
git commit -m "Add left-pad"
pnpm dlx turbo-ignore@canary docs --fallback=HEAD^1
```

**Comment by nathanhammond** [comment]
Tracing so far:
https://github.com/vercel/turbo/blob/60636fb4555f215abfbbfcb022f46f2907a7c979/cli/internal/scope/scope.go#L242-L248

In this if statement we presently exit with `makeAllPkgs()` which is ... wrong. We're at just `web` prior to this, so it's definitely calling back into Rust SCM land and failing there.

**Comment by nathanhammond** [comment]
Not SCM! It makes it through that, and to lockfile transitive closures. For `pnpm` it fails with `MissingResolvedVersion`:
`error performing lockfile operation: Unable to find resolved version for left-pad@^1.3.0 in apps/web`.

Worth noting, this only affects `pnpm`. I'm not able to reproduce for `npm`, `berry`, or `yarn`. So transitive closure calculation for `pnpm` has a bug.

**Comment by chris-olszewski** [comment]
This seems to be a bug I ported from Go. https://github.com/vercel/turbo/pull/6048 should fix it
